### PR TITLE
add: move-from-prod-to-dev-rev

### LINKED ISSUE #228: [FEATURE] Promote survey from "Development" to "Production" environment and vice versa
### Is your proposal related to a problem?

As a user, I have to create surveys twice: First in the development environment to test and then in the production environment. Having to set up surveys twice incl. logic is time-consuming and tedious, because I have to switch environments back and forth or be logged in in two browsers.

### Describe the solution you'd like

I'd like to have an option which says "Copy to Development" when I'm in the production environment and  "Copy to Production" when I'm in the development environment.

## Tasks

1. Add a dynamic item to the survey component in the SurveyList.tsx **(done)**
grafik

And in the production environment:
grafik

2. On click, copy the survey to the other environment (from prod to dev and vice versa).

3. Forward to the new environment

4. Display success toast (as drafted in the branch) **after** the user got forwarded to the new environment. 

## Hints
- There already is a "Duplicate" function in the SurveyList.tsx which you can use.
- There already is "Switching Environment" functionality in the rop right menu ( EnvironmentNavBar.tsx ) which you can use.
- I have added the UI in the branch "feature/copy-to-prod" which already adapts to the current environment.
- Make sure you only copy the plain survey and none of the responses or metadata (displays, etc.)
- Make sure to generate a new unique ID for the survey

### PR DESCRIPTION
**Pull request by Ankur-Datta-4**
## What does this PR do?

The change adds a feature where surveys could be moved from production environment to development environment, and vice-versa with just a button-click.
Fixes #228 (Feature)

 
[Loom video link](https://www.loom.com/share/6f3cf314bddb40f6a0b46a33099394a3?sid=07cdc1aa-e4ff-451f-ba71-1e36c1566067)

## Type of change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] Chore (refactoring code, technical debt, workflow improvements)
- [X] Enhancement (small improvements)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change adds a new database migration
- [ ] This change requires a documentation update

## How should this be tested?

This can be tested by choosing "move to prod env." in the drop-down of surveys in the dashboard (/environment/xx/surveys)

Tests which I ran:
- Move a survey from production to development env
- Move a survey from development to production env
- Ensure responses are not copied and only survey layout is copied

## Checklist

- [X] Read the [contributing guide](https://github.com/formbricks/formbricks/blob/main/CONTRIBUTING.md)
- [X] Self-reviewed my own code
- [X] Commented on my code in hard-to-understand bits
- [X] Ran `pnpm build`
- [X] Checked for warnings, there are none
- [X] Removed all `console.logs`
- [ ] Merged the latest changes from main onto my branch with `git pull origin main`
- [X] My changes don't cause any responsiveness issues
- [ ] Updated the Formbricks Docs if changes were necessary

### CODE REVIEW

**Comment by jobenjada** [comment]
@Ankur-Datta-4 works great, thanks!! 🙌 

@mattinannt pls review and merge

File: apps/web/app/environments/[environmentId]/surveys/SurveyList.tsx [hunk 1]
```diff
@@ -85,6 +94,22 @@ export default function SurveysList({ environmentId }) {
     }
   };
 
+  const copyToOtherEnvironment= async (surveyId)=>{
```
Comment (mattinannt) [inline apps/web/app/environments/[environmentId]/surveys/SurveyList.tsx:97]: Please install and use the `prettier` VSCode plugin to format your changes into the right format (spacing, etc.).

**Review by mattinannt — CHANGES_REQUESTED** [review]
@Ankur-Datta-4 Thanks for the PR 😊 I noticed a few things that need to be changed before we can merge the changes:

- Please use `Prettier` in VSCode to automatically format your code so that it looks consistent with the rest of the code in the repo.
- Please remove the period at the end of `Copy to Prod.`/`Copy to Dev.`.
- Please rename the copy of the survey and append `(copy)` to the name as we do when duplicating a survey.
- Currently the survey is copied without the triggers and audience settings. I discussed this with Johannes and we think that the expected behavior would be for the triggers and audience settings to be copied as well, so that you could, for example, use the exact survey in prod that you set up in dev. However, this comes with some added complexity: You will need to outsource your code to copy the survey to a custom api route so that you can use all the backend functionality. See the `duplicate` endpoint for more details ( https://github.com/formbricks/formbricks/blob/main/apps/web/pages/api/v1/environments/%5BenvironmentId%5D/surveys/%5BsurveyId%5D/duplicate.ts ). The attributeClasses and eventClasses may not already exist in this environment, so you'll need to check if an attributeClass / eventClass with the given name already exists or you'll need to create a new one and associate the survey with the new IDs.

I know the complexity of this ticket has increased significantly due to this change. Are you comfortable making these changes? 😊

**Comment by Ankur-Datta-4** [comment]
Sounds interesting, I'll work on duplicating the triggers & audience settings

**Comment by Ankur-Datta-4** [comment]
I've updated to make sure event-triggers are copied.
Here's a demo:

https://www.loom.com/share/c90b3e6a4bac48fa93b288674017c1fb?sid=00b6d321-a744-4192-b5b6-8aa0c767104b

Note:
- I have created a new route, modifying https://github.com/formbricks/formbricks/blob/main/apps/web/pages/api/v1/environments/%5BenvironmentId%5D/surveys/%5BsurveyId%5D/duplicate.ts 
- I've tested to ensure that: surveys get moved from dev-to-prod(or reverse) with/without triggers attached

File: apps/web/pages/api/v1/environments/[environmentId]/surveys/[surveyId]/duplicate/[targetEnvironmentId].ts [hunk 2]
```diff
@@ -0,0 +1,115 @@
...
+        // if the trigger does not exist in the target environment, create it
+        const newTrigger = await prisma.eventClass.create({
+          data: {
+            name: trigger.eventClass.name,
+            environment: {
+              connect: {
+                id: targetEnvironmentId,
+              },
+            },
+            createdAt: new Date(),
```
Comment (mattinannt) [inline apps/web/pages/api/v1/environments/[environmentId]/surveys/[surveyId]/duplicate/[targetEnvironmentId].ts:68]: `createdAt` and `updatedAt` are automatically set by prisma when creating a new `eventClass` and don't need to be set explicitly.

**Review by mattinannt — COMMENTED** [review]
Thanks for the PR 💪 The code looks great and the functionality works 🎉😊.
Unfortunately we already had missing functionality in the existing duplicate survey functionality that `SurveyAttributeFilters` doesn't copy over. Would you be willing to add this functionality to the existing duplicate endpoint as well as the new `duplicate/[targetEnvironmentId]` endpoint?

**Comment by mattinannt** [comment]
@Ankur-Datta-4 Thank you for adding the attributeFilters to the duplicate endpoint. Looks great!
Can you please also add this functionality to your new `.../duplicate/[targetEnvironmentId].ts` endpoint? 😊
I hope I'm not too impatient; if you're already working on it, don't let me pressure you 😉

**Comment by Ankur-Datta-4** [comment]
Yea working on it...facing some issues while testing. Will push it in a couple of hours

**Comment by Ankur-Datta-4** [comment]
I've followed the similar code (as done for events)
I've tested the same with previous quoted procedures.

File: apps/web/pages/api/v1/environments/[environmentId]/surveys/[surveyId]/duplicate/[targetEnvironmentId].ts [hunk 3]
```diff
@@ -0,0 +1,159 @@
...
+        targetEnvironmentTriggers.push(newTrigger.id);
+      } else {
+        targetEnvironmentTriggers.push(targetEnvironmentTrigger.id);
+      }
+    }
+
+
+    let targetEnvironmentAttributeFilters: string[] = [];
+    // map the local attributeFilters to the target env
+    for (let i = 0; i < existingSurvey.attributeFilters.length; i++) {
```
Comment (mattinannt) [inline apps/web/pages/api/v1/environments/[environmentId]/surveys/[surveyId]/duplicate/[targetEnvironmentId].ts:91]: a `for...of` loop is much easier to use for this functionality 😊
you can just write
```ts
for (const attributeFilter of existingSurvey.attributeFilters) {
...
}
```

File: apps/web/pages/api/v1/environments/[environmentId]/surveys/[surveyId]/duplicate/[targetEnvironmentId].ts [hunk 4]
```diff
@@ -0,0 +1,159 @@
...
+        where: {
+          name: attributeFilter.attributeClass.name,
+          environment: {
+            id: targetEnvironmentId,
+          },
+        },
+      });
+      if (!targetEnvironmentAttributeClass) {
+        const newAttributeClass=await prisma.attributeClass.create({
+          data:{
```
Comment (mattinannt) [inline apps/web/pages/api/v1/environments/[environmentId]/surveys/[surveyId]/duplicate/[targetEnvironmentId].ts:105]: It looks like your code isn't automatically run through prettier. We use prettier to always format the code automatically in the same style. You can find prettier for VSCode here: https://marketplace.visualstudio.com/items?itemName=esbenp.prettier-vscode
It's also recommended to use it as default formatter and format on save so every time you save a file, it automatically formats the file according to the project rules 😊

**Review by mattinannt — CHANGES_REQUESTED** [review]
@Ankur-Datta-4 Thanks for the changes 💪🤗 Functionality-wise works great, just had a few suggestions for the code and would be great if you could change them.
Code formatting / prettier is the most important part of this to keep a consistent code style. As an alternative to using prettier, you can also run `pnpm format` in the project root; but using prettier is highly recommended 😊.

When these changes are in we are ready to merge into main 🎉💪

**Comment by Ankur-Datta-4** [comment]
I've done the change. Today, I'm facing some issues with prettier like: 😌
image

**Comment by mattinannt** [comment]
> I've done the change. Today, I'm facing some issues with prettier like: 😌 image
> 
> Edit: pnpm format resulted in 603 changes. Im sure thats not what's required 😀

Interesting ;-) 
Can you share a screenshot of the files that were changed when you run `pnpm format`?
In my installation only changes to the `/duplicate/[targetEnvironmentId].ts` endpoint are made.
Would be good to know if different contributors have different `pnpm format` behaviour.

Screenshot 2023-06-21 at 16 00 36

**Comment by Ankur-Datta-4** [comment]
I see that it has modified every file there is! 😀
image

Every file from formbricks-com, web, demo, lib and database files

**Review by mattinannt — APPROVED** [review]
@Ankur-Datta-4 I changed the formatting in the last file and will merge the PR now 💪
Thanks a lot for this feature; awesome work 😊🤗

We will keep an eye on the problems with the `pnpm format` command, as I was unable to reproduce it at the moment.

**Pull request merged**